Add support for RISC-V architecture

bitsk opened the pull request:
<!--
Thank you for your pull request.
Below are a few things we ask you kindly to self-check before getting a review. Remove checks that are not relevant.
-->
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

<!--
Please note any issues this fixes using [closing keywords]( https://help.github.com/articles/closing-issues-using-keywords/ ):
-->

<!--
Please add any other relevant info below:
-->
This PR depends on merging [linux-sysroot-feedstock#98](https://github.com/conda-forge/linux-sysroot-feedstock/pull/98) first to resolve the RISC-V ABI directory issue. The GCC-13 build failed because the RISC-V `sysroot-with-crypt` is missing, as discussed https://github.com/conda-forge/ctng-compilers-feedstock/pull/127#discussion_r1573853024.

bitsk committed:
Add support for RISC-V architecture

bitsk commented:
@conda-forge-admin, please rerender

h-vetinari commented:
> The GCC-13 build failed because the RISC-V `sysroot-with-crypt` is missing, as discussed [here](https://github.com/conda-forge/ctng-compilers-feedstock/pull/127%23discussion_r1573853024).

That link is dead; however, there's no reason to build GCC 13 anymore[^1]. Just add a `skip: true  # [gcc_maj_ver == 13]`

[^1]: we're already on GCC 14 across conda-forge; that it's still built here is to keep things working for feedstocks that haven't transitioned to GCC 14 yet, but that's a very small number and for new architectures we're not constrained by any old builds

h-vetinari committed:
cross-compile riscv from x64

bitsk committed:
Bumped the build number

bitsk committed:
Skip build GCC13 for linux-riscv64

bitsk committed:
MNT: Re-rendered with conda-smithy 3.54.1 and conda-forge-pinning 2025.12.25.01.57.59

h-vetinari committed:
MNT: Re-rendered with conda-smithy 3.54.1 and conda-forge-pinning 2025.12.25.01.57.59

Other tools:
- conda-build 25.11.1
- rattler-build 0.54.0
- rattler-build-conda-compat 1.4.10

isuruf commented on recipe/build.sh:
```diff
@@ -140,6 +140,32 @@ if [[ ! ("${BUILD}" == "${HOST}" && "${HOST}" != "${TARGET}") && "${TARGET}" !=
+  # and when we switch to version 15 or later, 
+  # we can automatically switch to RVA23U64 and gradually replace the existing packages.
+  # The 'RVA23 Profiles' specification is available in
+  # https://drive.google.com/file/d/12QKRm92cLcEk8-5J9NI91m0fAQOxqNAq/view
```
I don't think we'd want to build for RVA23U64 even in gcc 15. Those are new (non-existing) hardware and we don't want to keep using gcc 14 just to build for those new hardware. The way we do this in x86_64 is to build gcc for the baseline and use CFLAGS, CXXFLAGS etc when building specific package.

bitsk replied:
Using CFLAGS and CXXFLAGS might not be suitable for all packages and could significantly increase the maintenance workload. It’s a bit of a dilemma at this point because RVA23 hardware is expected to hit the market in bulk by 2026.

We currently have a few options:

Build only rv64gc packages: This supports legacy hardware but lacks Bit-manipulation and Vector extensions (which are mandatory in aarch64), impacting performance for many packages.

Dual builds for both rv64gc and RVA23: This would double the required CI/build resources.

Gradual migration: Start with rv64gc and progressively transition to RVA23 (without a long-term commitment to legacy hardware).

Build RVA23 only: Focus exclusively on the new standard.

In short, we believe rv64gc-only hardware won't be around for the long term, and our distribution is already bottlenecked by its performance limitations. Given that new RISC-V chips will adopt the more mature RVA23 profile, I believe our RISC-V migration will align perfectly with the arrival of RVA23 hardware.

Which approach do you all prefer?

isuruf replied:
Build only rv64gc packages, but build RVA23 in specialized packages like OpenBLAS which can use runtime CPU feature detection.

h-vetinari replied:
IOW, I don't want to end up in a situation where our longterm baseline is rv64gc; we need to be able to stay up-to-date and relevant with our packaging. If rv64gc will quickly become obsolete as @bitsk predicts, then we need to be able to move on from that.

Given that we currently have _**zero**_ users on riscv, I don't think it's worth our time to start with rv64gc only to remigrate to RVA23 after the initial riscv migration. But if people prefer that, I won't stand in the way.

bitsk commented on recipe/meta.yaml:
```diff
@@ -222,7 +223,7 @@ outputs:
         - CC=$(${PREFIX}/bin/*-gcc -dumpmachine)-gcc
         - ${CC} -Wall tests/aligned_alloc.c -c -o c_aligned.o -v -Wl,-v -march=native  # [target_platform == cross_target_platform and (x86_64 or s390x)]
```
RISC-V does not support the `-march=native` option.

isuruf replied:
RISC-V doesn't support `-mcpu=native` either?

bitsk replied:
Yes，RISC-V GCC doesn‘t support ` -mcpu=native`，See https://gcc.gnu.org/onlinedocs/gcc/RISC-V-Options.html

isuruf commented:
>  is there a need for a broader discussion on the RISC-V baseline policy?

There should be. Targeting a baseline just for the OpenKylin community who has new chips seems like a recipe for disaster. We need buy-in from a larger community of users and limiting ourselves to one company's chips is bad IMO.

bitsk committed:
Modify --with-arch=rv64gc for GCC 15

h-vetinari commented on recipe/build.sh:
```diff
@@ -141,25 +141,12 @@ if [[ ! ("${BUILD}" == "${HOST}" && "${HOST}" != "${TARGET}") && "${TARGET}" !=
-    # Normalize whitespace then replace spaces with underscores: "i m a ..." -> "i_m_a_..."
-    RV_EXT="$(echo "${RV_EXT}" | tr -s '[:space:]' ' ' | sed -e 's/^ *//' -e 's/ *$//')"
-    GCC_CONFIGURE_OPTIONS+=(--with-arch="rv64${RV_EXT// /_}")
+  # According to discussions with core members, https://hackmd.io/@conda-community/SJ_oFitUZg
```
Here's a more stable link to the minutes if you'd like:
https://github.com/conda/governance/blob/main/meetings/archive/20260204-conda-ecosystem.md

bitsk committed:
Update RISC-V architecture references and remove obsolete patch

h-vetinari reviewed (approved):
Thanks, this LGTM. @isuruf, could you help unblock this please?

isuruf commented on recipe/build.sh:
```diff
@@ -140,6 +140,19 @@ if [[ ! ("${BUILD}" == "${HOST}" && "${HOST}" != "${TARGET}") && "${TARGET}" !=
+if [[ "$TARGET" == *riscv64* ]]; then
+  # According to discussions with core members, https://github.com/conda/governance/blob/main/meetings/archive/20260204-conda-ecosystem.md
+  # we have decided to continue using the rv64gc architecture in GCC 15, 
+  # and switch to the rva23u64 architecture in GCC 16 and later.
```
Can you add the condition of rocky moving to `rva23u64` here?

h-vetinari commented:
Hey @bitsk, could you address the remaining review feedback here please? I suggest to leave out discussion and configuration of rva23 completely for now. We can continue with this topic in https://github.com/conda-forge/conda-forge.github.io/issues/1744 or a separate issue, rather than holding up this PR (this discussion ends up costing weeks that we could have been spending on building packages already).

I want us to get to actually kick off the migration and start building stuff, which will uncover a bunch of work that needs doing anyway, regardless of `rv64gc` vs `rva23`. Doing a rebuild later for a different baseline will then be only a fraction of the work compared to the initial roll-out.

bitsk committed:
Update build script for RISC-V architecture and refine build skip conditions for GCC versions

bitsk commented:
@isuruf @h-vetinari  I've made the change, please take a look.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict. Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the `@` notation in a comment) if you believe this is a bug.

h-vetinari committed:
Merge branch 'main' into main

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I just wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found it was in an excellent condition.

I do have some suggestions for making it better though...


For **recipe/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.
 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/22622714334](https://github.com/conda-forge/conda-forge-webservices/actions/runs/22622714334). Examine the logs at this URL for more detail.</sub>

h-vetinari committed:
alphabetize conda-forge.yml

h-vetinari committed:
use alma10 images

h-vetinari committed:
MNT: Re-rendered with conda-smithy 3.55.1 and conda-forge-pinning 2026.03.03.10.06.51

Other tools:
- conda-build 26.1.0
- rattler-build 0.57.2
- rattler-build-conda-compat 1.4.11

h-vetinari commented:
Not sure how this worked before, but we're not running into issues with [qemu](https://github.com/conda-forge/conda-forge-ci-setup-feedstock/blob/1fcb6141c7b8979c53d857117539280aff57591f/recipe/cross_compile_support.sh#L65) in the ci-setup

```
+++ /usr/bin/qemu-riscv64-static --version
/opt/conda/bin/cross_compile_support.sh: line 65: /usr/bin/qemu-riscv64-static: No such file or directory
```

Looks like we need to expand https://github.com/conda-forge/docker-images/blob/main/download-qemu-static.sh (or put work into https://github.com/conda-forge/docker-images/issues/320)

bitsk commented:
We need to merge this PR first.https://github.com/conda-forge/linux-sysroot-feedstock/pull/98

h-vetinari commented:
> We need to merge this PR first.[conda-forge/linux-sysroot-feedstock#98](https://github.com/conda-forge/linux-sysroot-feedstock/pull/98)

That PR has been merged now, but we're still missing qemu support for riscv

isuruf commented:
@conda-forge-admin rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you but ran into some issues. Please check the [output logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/24095888861) of the GitHub Actions workflow below for more details. You can also ping `conda-forge/core` (using the `@` notation) for further assistance or you can try [rerendering locally](https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-with-conda-smithy-locally).


The following suggestions might help debug any issues:
* Is the `recipe/{meta.yaml,recipe.yaml}` file valid?
* If there is a `recipe/conda-build-config.yaml` file in the feedstock make sure that it is compatible with the current [global pinnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).
* Is the fork used for this PR on an organization or user GitHub account? Automated rerendering via the webservices admin bot only works for user GitHub accounts.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/24095888861](https://github.com/conda-forge/conda-forge-webservices/actions/runs/24095888861). Examine the logs at this URL for more detail.</sub>

bitsk committed:
Temporarily disable riscvv64 native package building

bitsk commented:
@conda-forge-admin rerender

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ The feedstock-name in the `extra` section must not end with '-feedstock'. The '-feedstock' suffix is automatically appended during feedstock creation.

For **recipe/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.
 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/24442666130](https://github.com/conda-forge/conda-forge-webservices/actions/runs/24442666130). Examine the logs at this URL for more detail.</sub>

bitsk commented:
Hi, @isuruf, Let's hold off on building the native riscv64 toolchain for now and submit the cross-compilation toolchain first. We can compile the native toolchain after its dependencies have been successfully cross-compiled. I have already added the new package names to the admin-request https://github.com/conda-forge/admin-requests/pull/2019.

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-webservice.

I tried to rerender for you but ran into some issues. Please check the [output logs](https://github.com/conda-forge/conda-forge-webservices/actions/runs/24442705283) of the GitHub Actions workflow below for more details. You can also ping `conda-forge/core` (using the `@` notation) for further assistance or you can try [rerendering locally](https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-with-conda-smithy-locally).


The following suggestions might help debug any issues:
* Is the `recipe/{meta.yaml,recipe.yaml}` file valid?
* If there is a `recipe/conda-build-config.yaml` file in the feedstock make sure that it is compatible with the current [global pinnings](https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/master/recipe/conda_build_config.yaml).
* Is the fork used for this PR on an organization or user GitHub account? Automated rerendering via the webservices admin bot only works for user GitHub accounts.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/24442705283](https://github.com/conda-forge/conda-forge-webservices/actions/runs/24442705283). Examine the logs at this URL for more detail.</sub>

isuruf commented:
> Let's hold off on building the native riscv64 toolchain for now and submit the cross-compilation toolchain first.

Can you skip the native toolchains here and rerender?

bitsk committed:
MNT: Re-rendered with conda-smithy 3.61.1 and conda-forge-pinning 2026.04.21.11.53.26

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I was trying to look for recipes to lint for you, but it appears we have a merge conflict. Please try to merge or rebase with the base branch to resolve this conflict.

Please ping the 'conda-forge/core' team (using the `@` notation in a comment) if you believe this is a bug.

bitsk committed:
Merge remote-tracking branch 'upstream/main'

conda-forge-admin commented:
Hi! This is the friendly automated conda-forge-linting service.

I wanted to let you know that I linted all conda-recipes in your PR (```recipe/meta.yaml```) and found some lint.

Here's what I've got...


For **recipe/meta.yaml**:

 * ❌ The feedstock-name in the `extra` section must not end with '-feedstock'. The '-feedstock' suffix is automatically appended during feedstock creation.
 * ❌ In conda-forge.yml: [`$.azure`](https://conda-forge.org/docs/maintainer/conda_forge_yml/#azure) `=` `{'free_disk_space': True, 'store_build_artifacts': True}`.
    > {'free_disk_space': True, 'store_build_artifacts': True} is not valid under any of the given schemas
    <details>
    <summary>Schema</summary>

    ```json
    {
      "anyOf": [
        {
          "$ref": "#/$defs/AzureConfig"
        },
        {
          "type": "null"
        }
      ]
    }
    ```

    </details>

For **recipe/meta.yaml**:

 * ℹ️ The recipe is not parsable by parser `conda-souschef (grayskull)`. This parser is not currently used by conda-forge, but may be in the future. We are collecting information to see which recipes are compatible with grayskull.
 * ℹ️ The recipe is not parsable by parser `conda-recipe-manager`. The recipe can only be automatically migrated to the new v1 format if it is parseable by conda-recipe-manager.


<sub>This message was generated by GitHub Actions workflow run [https://github.com/conda-forge/conda-forge-webservices/actions/runs/24825473556](https://github.com/conda-forge/conda-forge-webservices/actions/runs/24825473556). Examine the logs at this URL for more detail.</sub>
